mopro-ffi/src/halo2/mod.rs

```diff
@@ -1,69 +1,172 @@
-    //     bincode::deserialize(&public_inputs).map_err(|e| MoproError::Halo2Error(e.to_string()))?;
-    // let is_valid = halo2::verify_halo2_proof(deserialized_proof, deserialized_inputs).unwrap();
-    // Ok(is_valid)
+/// Proving Circuits are Halo2 Circuits that can generate proofs
+/// Provide the circuits that you want to be able to generate proofs for
+/// as a list of pairs of the form `circuit_proving_key`, `prove_fn`
+/// Where `circuit_proving_key` is the name of the proving key file
+/// and `prove_fn` is the function that generates the proof available locally
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/halo2` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_halo2_proving_circuits {
+    // Generates a function `set_circom_circuits` that takes no arguments and updates CIRCOM_CIRCUITS
+    ($($key:expr, $func:expr),+ $(,)?) => {
+        fn set_halo2_proving_circuits() -> std::collections::HashMap<String, mopro_ffi::Halo2ProveFn> {
+            let mut circuits: std::collections::HashMap<String, mopro_ffi::Halo2ProveFn> = std::collections::HashMap::new();
+
+            $(
+                    circuits.insert($key.to_string(), $func);
+            )+
+
+            circuits
+        }
+    };
 }
 
-#[test]
-fn test_end_to_end() -> Result<(), MoproError> {
-    // We by default compile the Fibonacci Halo2 Circuit
-    // TODO - For the future we should consider a stateful circuit to change the keys on the fly.
-
-    let mut inputs = HashMap::new();
-    let out = 55u64;
-    inputs.insert("out".to_string(), vec![out.to_string()]);
-
-    let expected_output = vec![Fr::from(1), Fr::from(1), Fr::from(out)];
-    let expected_output_bytes = bincode::serialize(&SerializablePublicInputs(expected_output))
-        .expect("Serialization of Output Expected bytes failed");
-
-    // Step 2: Generate Proof
-    let generate_proof_result = generate_halo2_proof(inputs)?;
-    let serialized_proof = generate_proof_result.proof;
-    let serialized_inputs = generate_proof_result.inputs;
-
-    assert!(serialized_proof.len() > 0);
-    assert_eq!(serialized_inputs, expected_output_bytes);
-
-    // Step 3: Verify Proof
-    let is_valid = verify_halo2_proof(serialized_proof.clone(), serialized_inputs.clone())?;
-    assert!(is_valid);
+/// Verifying Circuits are Halo2 Circuits that can verify proofs
+/// Provide the circuits that you want to be able to verify proofs for
+/// as a list of pairs of the form `circuit_verifying_key`, `verify_fn`
+/// Where `circuit_verifying_key` is the name of the verifying key file
+/// and `verify_fn` is the function that verifies the proof available locally
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/halo2` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_halo2_verifying_circuits {
```

Comment:
Let's try to combine the prove/verify mapping. It should be possible to return a tuple of functions.

Reply:
I did combine them under the same macro (the simple approach):
```rust
mopro_ffi::set_halo2_circuits! {
    ("fibonacci_pk", halo2_fibonacci::prove, "fibonacci_vk", halo2_fibonacci::verify)
}
```

However, under the hood they are still different functions, to serve the advanced users who might want to have more tailored logic:

 As effectively a user can select if they want to have a halo2 functionality for only verifying or also proving. And that implies that they should have an option to only specify the circuit they want to be able to verify, while avoiding adding any proving circuits. By merging it all into a single function that removes the flexibility, which will likely be very important for the users.